[Connector]Add plugin for Apache Pulsar.

MarvinCai opened the pull request:
Fix #7852
Add a plugin for Apache [Pulsar](https://pulsar.apache.org/).
It's mostly a migration of original Pulsar PrestoSQL [connector](https://github.com/apache/pulsar/tree/master/pulsar-sql/presto-pulsar) with update for all the package to Trino and fix dependency and code style issues.
This plugin only support read from Pulsar for now, write capability will be added later.

kokosing commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,514 @@
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
```
What is the point of shading here? Connectors typically are not used as libraries.

MarvinCai replied:
It was used as a separate jar before, I guess we don't need it here, removed.

kokosing commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,514 @@
+            <groupId>org.apache.bookkeeper</groupId>
+            <artifactId>bookkeeper-server</artifactId>
+            <version>${bookkeeer.version}</version>
+            <exclusions>
```
Why that many deps are excluded? If you hit a dependency conflict, then exclusion is not a proper fix of this issue as using classes from dependency in different version may not work as expected. You might want to shade these dependencies like https://github.com/trinodb/trino-hadoop-apache. I suggest to shade all deps you need for pulsar connector in one project.

MarvinCai replied:
Found out we have a shaded version, switched to that.
Added some boilerplate code to convert between shaded version of Netty ByteBuf and normal version.
Also still need to ping some Dep to specific version like Netty as we used some classes from higher version.

kokosing commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,514 @@
+
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
```
We are not using mocks. See https://github.com/trinodb/trino/blob/master/DEVELOPMENT.md#avoid-mocks. We don't believe that they improve test coverage, but surely they add a lot of complexity and maintenance cost.

MarvinCai replied:
Switch from mockito to mock classes, removed mockito Dep.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnHandle.java:
```diff
@@ -0,0 +1,267 @@
+
+    /**
+     * Column Name.
+     */
```
Such comments are pointless.

MarvinCai replied:
agree, removed.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnector.java:
```diff
@@ -0,0 +1,112 @@
+    public final void shutdown()
+    {
+        try {
+            this.pulsarConnectorConfig.close();
```
not needed `this.` prefix (please remove it from other places too).

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnector.java:
```diff
@@ -0,0 +1,112 @@
+        try {
+            this.pulsarConnectorConfig.close();
+        }
+        catch (Exception e) {
```
Catching plain `Exception` is tricky as you would need to handle `InterruptedException` too. Here I would suggest to you not to catch anything, or use `Closer` or use `Closeables.closeAll` or `try-with-resources`.

MarvinCai replied:
you're right, end up just removing the try-catch.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCache.java:
```diff
@@ -0,0 +1,213 @@
+import org.apache.bookkeeper.mledger.offload.Offloaders;
+import org.apache.bookkeeper.mledger.offload.OffloadersCache;
+import org.apache.bookkeeper.stats.StatsProvider;
+import org.apache.commons.lang3.StringUtils;
```
Prefer guava over apache commons lang3

MarvinCai replied:
Switched from lang3.ExceptionUtils to guava Throwables
lang3.Pair to AbstractMap.SimpleEntry
lang3.StringUtils to guava Strings

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCache.java:
```diff
@@ -0,0 +1,213 @@
+                }
+            }
+            else {
+                log.info("No ledger offloader configured, using NULL instance");
```
most of such logging should be on debug. How much is that helpful for users?

MarvinCai replied:
Changed all of them to log.debug with check for log.isDebugEnabled, and removed some which are just showing result of different configuration.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCache.java:
```diff
@@ -0,0 +1,213 @@
+                return NullLedgerOffloader.INSTANCE;
+            }
+        }
+        catch (Throwable t) {
```
Please do not catch `Throwable` that may include OOM, or interrupted exceptions.

kokosing reviewed (commented):
Few initial commits. Please check why the build is failing. It looks like the license header you are using is different than expected. But there might be other formatting issues as well.

kokosing commented:
@MarvinCai If you any of the comment is not clear please reach out to me on slack. https://trino.io/slack.html

MarvinCai commented:
@kokosing 
Sorry for the delay, took some time to replace usage of Mockito with mock classes.
And fix the license header, didn't notice the content was actually different, just thought they're the same Apache 2.0 license.
Please take another look.

kokosing commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+=================
```
@mosabua Can you please take a look?

kokosing commented on .github/workflows/ci.yml:
```diff
@@ -216,6 +216,7 @@ jobs:
             !:trino-hive,!:trino-orc,!:trino-parquet,
             !:trino-mongodb,!:trino-kafka,!:trino-elasticsearch,
             !:trino-redis,
+            !:trino-pulsar,
```
Let's run pulsar together with redis or with kafka above. Your build takes just 4 minutes so I guess compilation takes the most.

MarvinCai replied:
merged with Redis test, it'll take longer as BaseConnectorTest is added, but we can split later if needed.

kokosing commented on .github/workflows/ci.yml:
```diff
@@ -216,6 +216,7 @@ jobs:
             !:trino-hive,!:trino-orc,!:trino-parquet,
```
Let's squash all commits. I suggest to use below for commit message:
```
Introduce connector for Apache Pulsar
```

MarvinCai replied:
hmmmm, I thought all commits will be squashed to one commit when we merge the pr? is there anything I need to do here?

mosabua replied:
you need to rebase them into one commit yourself and then force push into the branch

kokosing commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+.. code-block:: text
+
+    connector.name=pulsar
+    pulsar.broker-service-url=http://localhost:8080
```
localhost? It looks like trino discover uri. Maybe let's use some placeholder instead.

bitsondatadev replied:
It looks like this is actually the rest port used by Pulsar. We will need to think of how to deal with this. I think traditionally Trino was an internally managed service that shipped with Pulsar so they likely gave Trino different ports. If we assume a Trino centric point of view, we would likely want to change the port of Pulsar.

https://pulsar.apache.org/docs/en/next/standalone-docker/#use-pulsar-in-docker

MarvinCai replied:
this is actually the url of Pulsar Admin endpoint, and default value will most likely used in standalone deployment which is used for experimenting/exploring. Will specify in doc that user should change to real url when used in non-test deployment.

kokosing commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+    pulsar.managed-ledger-num-worker-threads =
+    pulsar.managed-ledger-num-scheduler-threads =
+
+    pulsar.stats-provider-configs={"httpServerEnabled":"false", "prometheusStatsHttpPort":"9092", "prometheusStatsHttpEnable":"true"}
```
Does user has to configure that all? Maybe we should expose the minimal viable configuration and leave others as advanced configuration. In general connector should have good defaults, so it can run in most cases without additional configuration.

MarvinCai replied:
it's actually not a required config, will remove from sample config file.

kokosing commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+The following configuration properties are available:
+
+================================================      =======================================================================
+Property Name                                         Description
```
Please mention default values (as separate column)?

I guess this format for table is obsolete and we use something else. @mosabua any ideas?

MarvinCai replied:
will update.

kokosing commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+``pulsar.nar-extraction-directory``                   Directory to use for extraction Nar file
+================================================      =======================================================================
+
+``pulsar.broker-service-url``
```
Do we need to repeat this? It is already above.

mosabua replied:
Lets not use the following setup. I would suggest to work all the content below into the table above. Convert it to use the list-table setup though and stay with the 80 char column hard wrap

mosabua replied:
Even better would be to break the table and config section up into multiple sections for example about required config, bookkeeper config, security config and others and have separate tables. Each section could then have more info to explain details for this topic.

Anonymitaet replied:
@mosabua good suggestion! We've updated.

kokosing commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,274 @@
+        <air.check.skip-dependency>true</air.check.skip-dependency>
+        <air.check.fail-checkstyle>false</air.check.fail-checkstyle>
+        <air.check.skip-duplicate-finder>true</air.check.skip-duplicate-finder>
+        <air.check.skip-modernizer>true</air.check.skip-modernizer>
```
Please try to enable all of these airbase related properties. For example `air.check.skip-duplicate-finder` skip is dangerous as it points the issues with dependencies that could manifest on production. That would be unfortunate.

MarvinCai replied:
Spent a lot of time try to solve the dependency issue, tried couple shaded version but didn't work well. (It depends on un-shade dependency and will cause NoSuchMethod at run time.)
Currently still need to skip dependency check and duplicate check, 
As Pulsar is using dep in Netty 4.1.63 and need to pin to this version, which will fail dependency check.
And some Jakarta/Javax library will cause duplicate to fail.
But I tried to run the connector with latest Trino version and seems working fine.

kokosing commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,274 @@
+        </dependency>
+
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
```
Is it possible to use jackson instead?

MarvinCai replied:
removed this dep.

kokosing commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,274 @@
+        </dependency>
+
+        <dependency>
+            <groupId>commons-logging</groupId>
```
Please use airlift-log

MarvinCai replied:
removed this dep.

kokosing commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,274 @@
+
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
```
Is it possible to remove this?

MarvinCai replied:
removed.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnHandle.java:
```diff
@@ -0,0 +1,262 @@
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class represents the basic information about a trino column.
```
Pointless comment. Please remove comments like that.

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnHandle.java:
```diff
@@ -0,0 +1,262 @@
+            @JsonProperty("mapping") String mapping,
+            @JsonProperty("dataFormat") String dataFormat,
+            @JsonProperty("formatHint") String formatHint,
+            @JsonProperty("handleKeyValueType") HandleKeyValueType handleKeyValueType)
```
Optional?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnMetadata.java:
```diff
@@ -0,0 +1,233 @@
+
+    private DecoderExtraInfo decoderExtraInfo;
+
+    public PulsarColumnMetadata(String name, Type type, String comment, String extraInfo,
```
We usually place each argument in separate line. Please follow that here and everywhere else.

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnector.java:
```diff
@@ -0,0 +1,96 @@
+    @Override
+    public ConnectorMetadata getMetadata(ConnectorTransactionHandle transactionHandle)
+    {
+        return new ClassLoaderSafeConnectorMetadata(metadata, getClass().getClassLoader());
```
Wrapping with `ClassLoaderSafeConnectorMetadata` should happen in guice module.

MarvinCai replied:
Update to bind and inject `ClassLoaderSafeConnectorMetadata` with `ConnectorMetadata` for PulsarPlugin.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnector.java:
```diff
@@ -0,0 +1,96 @@
+    private final PulsarMetadata metadata;
+    private final PulsarSplitManager splitManager;
+    private final PulsarRecordSetProvider recordSetProvider;
+    private final PulsarConnectorConfig pulsarConnectorConfig;
```
Can you please depend on generic type like `ConnectorRecordSetProvider`?

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnector.java:
```diff
@@ -0,0 +1,96 @@
+    @Override
+    public final void shutdown()
+    {
+        pulsarConnectorConfig.close();
```
Config class should be plain beans, they should not be closeable. I think you wanted to have a separate entity that could be closed. 

Also you don't have to do it here. You can use `@PostConstruct` annotation.

MarvinCai replied:
It's a Admin client talking to Pulsar used to get metadata for topic, we only want to close it when we're done using the connector, so removed it to a separate class but still keep its `close` method in `shutdown`.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCacheImpl.java:
```diff
@@ -0,0 +1,208 @@
+    private static final Logger log = Logger.get(PulsarConnectorCacheImpl.class);
+
+    @VisibleForTesting
+    static PulsarConnectorCache instance;
```
This bad design. It is not going to work if user to have several pulsar catalogs. Currently we have classloader separation that each plugin has its own plugin, but catalogs that are using same connector (same plugin then) will share the same classloader. In same cases we go even further and we can duplicate classloader per each catalog, but still this design is bad.

Please use guice to inject that instance.

MarvinCai replied:
Sure thing, changed to use Guice to inject it.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCacheImpl.java:
```diff
@@ -0,0 +1,208 @@
+                            getOffloaderScheduler(offloadPolicies));
+                }
+                catch (IOException ioe) {
+                    log.error("Failed to create offloader: ", ioe);
```
no point to log if you throw. Please fix other places like that too.

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCacheImpl.java:
```diff
@@ -0,0 +1,208 @@
+                    throw new RuntimeException(ioe.getMessage(), ioe.getCause());
+                }
+            }
+            else {
```
else block not needed

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestPulsarRecordCursor.java:
```diff
@@ -0,0 +1,496 @@
+import static org.testng.Assert.assertTrue;
+
+@Test(singleThreaded = true)
+public class TestPulsarRecordCursor
```
I don't really care about such tests. They are hard to maintain and do not help much. 

I would suggest to you implement `BaseConnectorTest` for pulsar. Please also see other connectors to see how you can test type mapping. We prefer to run entire query and see the results that user could see. That way functionality is covered and even if you change internals of product code, the tests are unchanged and they are up to date providing the coverage.

That is the reason why don't like excessive mocking, not that we don't like mockito.

MarvinCai replied:
Added a test based on `BaseConnectorTest`, disabled some test not applicable to Pulsar connector for now like create/drop table, or describe table as Pulsar table has some internal columns.

kokosing reviewed (commented):
Another round of initial comments.

mosabua commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+=================
+Pulsar connector
+=================
```
Empty line after title

mosabua commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+and the default number is 3. In Trino Pulsar connector, data is read directly from BookKeeper,
+so Trino workers can read concurrently from horizontally scalable number BookKeeper nodes.
+
+This connector is **read-only** for now. It can only read data from Pulsar topics,
```
This whole section should go into a "SQL support" section further down the file. Something like

```
SQL support
-----------

The connector provides **read-only access** to data and metadata in Pulsar topics
with the following statements:

* :doc:`/sql/select`
* :doc:`/sql/show-catalogs`
* :doc:`/sql/show-columns`
* :doc:`/sql/show-schemas`
* :doc:`/sql/show-tables`
* :doc:`/sql/describe`

Other supported commands:

* :doc:`/sql/explain`
* :doc:`/sql/explain-analyze`
* :doc:`/sql/set-session`
* :doc:`/sql/show-functions`
* :doc:`/sql/show-session`
* :doc:`/sql/use`
* :doc:`/sql/values`

No other statements are supported.
```

mosabua replied:
You might have to adapt what is true for this connector and add this section after the "Configuration" section

Anonymitaet replied:
@mosabua thanks for reminding, we will test and add docs.

mosabua commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+which enables concurrent reads and high read throughput. You can configure the number of BookKeeper nodes,
+and the default number is 3. In Trino Pulsar connector, data is read directly from BookKeeper,
+so Trino workers can read concurrently from horizontally scalable number BookKeeper nodes.
+
```
Probably here insert a Requirements section modelled in wording after e.g. the sqlserver.rst currently in master

Anonymitaet replied:
Yes, we've added the Requirements section in the draft and need your review later, thanks

mosabua commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+
+This connector is **read-only** for now. It can only read data from Pulsar topics,
+but cannot create topic or write to exiting topic yet.
+
```
Add a title here

Configuration

mosabua commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+This connector is **read-only** for now. It can only read data from Pulsar topics,
+but cannot create topic or write to exiting topic yet.
+
+To configure the Pulsar connector, create a catalog properties file ``etc/catalog/pulsar.properties``
```
Reword this so the catalog file name is NOT identical to the connector name. See for example sqlserver.rst and adapt to Pulsar

Anonymitaet replied:
@mosabua OK, we've changed it to `apachepulsar.properties`.

mosabua commented on docs/src/main/sphinx/connector/pulsar.rst:
```diff
@@ -0,0 +1,270 @@
+To configure the Pulsar connector, create a catalog properties file ``etc/catalog/pulsar.properties``
+A sample is provided:
+
+.. code-block:: text
```
Adapt this to the minimum required catalog file.

mosabua replied:
If necessary for specific use cases add separate complete config file examples but mostly rely on the table of properties.

Anonymitaet replied:
@mosabua OK, we've added minimum configuration.

mosabua reviewed (changes requested):
A bunch of doc structure and content changes needed. If desired I can help in a chat on slack or even a call for a quick sync and work together.

Anonymitaet commented:
Hi @kokosing @bitsondatadev @mosabua thank you very much for your advice on the docs! 
@MarvinCai and I are working on the docs and incorporating your comments https://github.com/trinodb/trino/issues/7852#issuecomment-852902162.
Feel free to ping us if you have more suggesitons.

MarvinCai commented:
@mosabua Thanks for reviewing the doc, we actually want to separate the doc and code so we'll have a separate pr for the doc and technical writer from out community @Anonymitaet will work on the doc, all the comments will be taken care of.
@kokosing Changed to depend on Pulsar 2.8 which should be released by end of next week it contains important fix(https://github.com/apache/pulsar/pull/9663) which will allow Pulsar connector to read compressed message, currently depending on 2.8-SNAPSHOT so some tests are failing, will fix before merge.
I added test based on `BaseConnectorTest`, it runs fine in my IDE, but it's weird that it'll fail when running from maven build due to somehow not able to connect to Pulsar/Zookeeper running in container, not sure if it can pass in github ci.

mosabua commented:
Sounds great @MarvinCai .. it would be good to chat with @Anonymitaet to make it more efficient for her. Please feel free to reach out on slack and I can arrange a meeting in which we could hack on it together.

Anonymitaet commented:
@mosabua thanks for your help. I would also like to have a meeting to discuss the [docs](https://github.com/trinodb/trino/issues/7852#issuecomment-862194785) if needed.

mosabua commented:
@Anonymitaet can you send me an email with your timezone and a few available time slots this week or next week .. maybe 1h to start.. then I can set up a meeting with video call to work on this for a bit

Anonymitaet commented:
@mosabua sure, I've sent an email to you and invited @MarvinCai to attend this meeting as well (as we might need tech inputs and discussions). Looking forward to meeting you guys.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/protobufnative/package-info.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * This package contains decoder for SchemaType.AVRO.
+ */
+package io.trino.plugin.pulsar.decoder.protobufnative;
```
We don't use this files.

MarvinCai replied:
removed

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/protobufnative/PulsarProtobufNativeRowDecoderFactory.java:
```diff
@@ -0,0 +1,169 @@
+        return dataType;
+    }
+
+    private static final Logger log = Logger.get(PulsarProtobufNativeRowDecoderFactory.class);
```
we define these loggers at the top of the class (convention)

MarvinCai replied:
updated

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/json/PulsarJsonFieldDecoder.java:
```diff
@@ -0,0 +1,480 @@
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Copy from {@link io.trino.decoder.json.DefaultJsonFieldDecoder} (trino-record-decoder)
```
Instead of copying, can we make DefaultJsonFieldDecoder extendable?

MarvinCai replied:
sure, made necessary change in DefaultJsonFieldDecoder to make it extendable and reduced code duplication in PulsarJsonFieldDecoder.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoder.java:
```diff
@@ -0,0 +1,79 @@
+            avroRecord = record.getAvroRecord();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
```
Please remove. And any other occurences of `printStackTrace` from production code.

MarvinCai replied:
removede

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoder.java:
```diff
@@ -0,0 +1,79 @@
+            GenericAvroRecord record = (GenericAvroRecord) genericAvroSchema.decode(byteBuf);
+            avroRecord = record.getAvroRecord();
+        }
+        catch (Exception e) {
```
This is too broad exception.

MarvinCai replied:
change exception scope to just SchemaSerializationException

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarTopicDescription.java:
```diff
@@ -0,0 +1,72 @@
+            @JsonProperty("topicName") String topicName)
+    {
+        checkArgument(!isNullOrEmpty(tableName), "tableName is null or is empty");
+        this.tableName = tableName;
```
Can `tableName` be null? If not please use `requireNonNull` for any of such fields that are non nullable.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarTransactionHandle.java:
```diff
@@ -0,0 +1,24 @@
+import io.trino.spi.connector.ConnectorTransactionHandle;
+
+/**
+ * A handle for transactions.
```
Please remove such comments that does not provide value.

MarvinCai replied:
removed

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSplitManager.java:
```diff
@@ -0,0 +1,501 @@
+
+    private final PulsarConnectorCache pulsarConnectorManagedLedgerFactory;
+
+    private static final Logger log = Logger.get(PulsarSplitManager.class);
```
static fields before normal

MarvinCai replied:
updated fields order

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSchemaInfoProvider.java:
```diff
@@ -0,0 +1,121 @@
+    {
+        ClassLoader originalContextLoader = Thread.currentThread().getContextClassLoader();
+        try {
+//            Thread.currentThread().setContextClassLoader(InjectionManagerFactory.class.getClassLoader());
```
remove

MarvinCai replied:
actually it's needed, I thought it was causing dependency issue when start up the service so comment it out for testing, I've removed the comment.

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestBasePulsarConnectorTest.java:
```diff
@@ -0,0 +1,116 @@
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        pulsarServer = new PulsarServer(PulsarServer.DEFAULT_IMAGE_NAME);
```
use `closeAfterTest`

MarvinCai replied:
I think you mean closeAfterClass(), updated to wrap the test server with closeAfterClass().

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestBasePulsarConnectorTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Override
+    public void testCreateTable()
```
This makes an impression that test is passing while it is not not. If feature is not supported it should fail nicely. For example with message that `Feature is not supported in Pulsar connector` and this should be tested here. Usually we use `hasBehavior` to define what to expect from the connector and `BaseConnectorTest` should have proper assertions for postive and negative tests. If you are missing `TestingConnectorBehavior` for given test please add it and improve the test if needed.

MarvinCai replied:
updated test so it now test the behavior properly.

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/PulsarQueryRunner.java:
```diff
@@ -0,0 +1,79 @@
+    {
+        Logging.initialize();
+
+        DistributedQueryRunner queryRunner = createPulsarQueryRunner(
```
Please start Trino on 8080

MarvinCai replied:
updated.

kokosing reviewed (commented):
Reading and learning. I will try to add more comments tomorrow.

mosabua commented on docs/src/main/sphinx/connector.rst:
```diff
@@ -31,6 +31,7 @@ from different data sources.
     Pinot           <connector/pinot>
```
This is already planned .. I think @MarvinCai just has to follow up in this PR on backing the change out. 

I am working with @Anonymitaet and @MarvinCai

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/package-info.java:
```diff
@@ -0,0 +1,14 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.pulsar.decoder.avro;
```
remove

MarvinCai replied:
removed

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroColumnDecoder.java:
```diff
@@ -0,0 +1,429 @@
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Copy from {@link io.trino.decoder.avro.AvroColumnDecoder} (trino-record-decoder)
```
Can improve AvroColumnDecoder to support the below types?

MarvinCai replied:
Since we now use a shaded dep, we're also using shaded version of avro here, it's hard to convert shaded version of avro to original avro class, I spent some time but was not able to make the conversion to make use of code in AvroColumnDecoder. Can we keep it as it for now and make the improve later?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoderFactory.java:
```diff
@@ -0,0 +1,196 @@
+                                    handleKeyValueType, new PulsarColumnMetadata.DecoderExtraInfo(field.name(),
+                                    null, null))
+
+                    ).collect(toList());
```
Move ) to previous line, collect to immutable list

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoderFactory.java:
```diff
@@ -0,0 +1,196 @@
+            log.warn(e, "Topic "
+                    + topicName.toString() + " extractColumnMetadata failed.");
+            throw new TrinoException(NOT_SUPPORTED, "Topic "
+                    + topicName.toString() + " schema may contains cyclic definitions.", e);
```
Please do not break lines

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoderFactory.java:
```diff
@@ -0,0 +1,196 @@
+        return columnMetadata;
+    }
+
+    private Type parseAvroTrinoType(String fieldName,
```
Please do not break lines like this. Either all parameters in single line or, all in separate (then fieldName needs to go subsequent line)

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoderFactory.java:
```diff
@@ -0,0 +1,196 @@
+                    ).collect(toList());
+        }
+        catch (StackOverflowError e) {
+            log.warn(e, "Topic "
```
No need to log anything as exception is propagated

MarvinCai replied:
removed.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/avro/PulsarAvroRowDecoderFactory.java:
```diff
@@ -0,0 +1,196 @@
+
+                    ).collect(toList());
+        }
+        catch (StackOverflowError e) {
```
Why it can happen?

MarvinCai replied:
If a avro schema contains a field referring itself, it could possibly cause stackoverflow when trying to extract column matadata from the schema. I think it's more like a safety net.

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestBasePulsarConnectorTest.java:
```diff
@@ -0,0 +1,116 @@
+import static io.trino.plugin.pulsar.PulsarServer.SELECT_FROM_ORDERS;
+import static io.trino.plugin.pulsar.PulsarServer.SELECT_FROM_REGION;
+
+public class TestBasePulsarConnectorTest
```
TestBasePulsarConnectorTest -> TestPulsarConnectorTest

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestPulsarSplitManager.java:
```diff
@@ -0,0 +1,449 @@
+import static org.testng.Assert.fail;
+
+@Test(singleThreaded = true)
+public class TestPulsarSplitManager
```
Please port these tests to TestBasePulsarConnectorTest. Tests here are too low level. While testing this on TestBasePulsarConnectorTest will show what is the user impact and what use case scenario is supported.

MarvinCai replied:
sure, ported necessary tests into TestPulsarConnectorTest, since TestPulsarConnectorTest is already testing many functionalities at a higher level, just removed these 3 test classes.

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestPulsarMetadata.java:
```diff
@@ -0,0 +1,400 @@
+import static org.testng.Assert.fail;
+
+@Test(singleThreaded = true)
+public class TestPulsarMetadata
```
Please port these tests to TestBasePulsarConnectorTest. Tests here are too low level. While testing this on TestBasePulsarConnectorTest will show what is the user impact and what use case scenario is supported.

kokosing reviewed (commented):
Still reading...

Anonymitaet commented:
Hi @MarvinCai I've submitted the doc PR https://github.com/trinodb/trino/pull/8469, you can revert doc changes in this PR.

Any progress on this code PR? thanks

MarvinCai commented:
addressed 1/3 of the comments, will address the rest soon

Anonymitaet commented:
@MarvinCai OK, look forward your changes.

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,306 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.12.2</version>
```
removed version for jackson

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,306 @@
+        <bookkeeer.version>4.14.1</bookkeeer.version>
+        <netty.version>4.1.63.Final</netty.version>
+        <air.check.skip-dependency>true</air.check.skip-dependency>
+        <air.check.skip-duplicate-finder>true</air.check.skip-duplicate-finder>
```
as discussed, created a new shaded jar for Trino Pulsar connector, I've tested locally and it works fine, if you think the updated deps looks good I'll send you the pom so you can help create a project for it.

MarvinCai commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarAdminClientProvider.java:
```diff
@@ -0,0 +1,69 @@
+
+public class PulsarAdminClientProvider
+{
+    private static PulsarAdmin pulsarAdmin;
```
I've tried to use a Provider to inject PulsarAdmin and it works fine, but seems in Guice document they mention we shouldn't use Guice to inject Closable resources: https://github.com/google/guice/wiki/Avoid-Injecting-Closable-Resources
I'm not sure if injecting PulsarAdmin without closing it here will cause any resource leak or not as it's mostly just a REST client wrapper to interact with REST endpoint to get meta info.

MarvinCai commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorConfig.java:
```diff
@@ -0,0 +1,511 @@
+public class PulsarConnectorConfig
+{
+    private String webServiceUrl = "http://localhost:8080";
+    private String zookeeperUri = "localhost:2181";
```
removed default and made it a required config.

Anonymitaet commented:
Hi @MarvinCai I see you've made code changes, do we need to update docs (https://github.com/trinodb/trino/pull/8469) accordingly? thanks

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarMetadata.java:
```diff
@@ -0,0 +1,438 @@
+    }
+
+    @Override
+    public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle)
```
Table layouts are going to be removed.  Can you please return `true` for `io.trino.spi.connector.ConnectorMetadata#usesLegacyTableLayouts`

MarvinCai replied:
updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarMetadata.java:
```diff
@@ -0,0 +1,438 @@
+            }
+            else if (e.getStatusCode() == 401) {
+                throw new TrinoException(QUERY_REJECTED,
+                        String.format("Failed to get pulsar topic schema information for topic %s: Unauthorized",
```
static imports for `format`

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarMetadata.java:
```diff
@@ -0,0 +1,438 @@
+                                topicName));
+            }
+            else {
+                throw new RuntimeException("Failed to get pulsar topic schema information for topic "
```
use format

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestPulsarConnectorTest.java:
```diff
@@ -0,0 +1,206 @@
+    protected PulsarServer pulsarServer;
+
+    @AfterClass(alwaysRun = true)
+    public void destroy()
```
This is not needed.

MarvinCai replied:
yeah, already wrapped with closeAfterClass, removed.

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestPulsarConnectorTest.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    @Override
+    protected QueryRunner createQueryRunner()
```
please move it to the top

MarvinCai replied:
done.

kokosing reviewed (commented):
I left few random comments... I need to get back to it. It looks much better.

MarvinCai commented:
thanks, I'll address them but hold new revision till you read through it.

MarvinCai commented on .github/workflows/ci.yml:
```diff
@@ -232,7 +232,7 @@ jobs:
             !:trino-hive,!:trino-orc,!:trino-parquet,
             !:trino-mongodb,!:trino-kafka,!:trino-elasticsearch,
-            !:trino-redis,
+            !:trino-redis,!:trino-pulsar,
```
yeah, previously there's only unit test so we thought it can be merged with other test, now that we added query test and it takes 10+ mins, probably better to run as a separate step. updated

kokosing replied:
Thanks, please make exclusion for trino-pulsar also in separate line.

MarvinCai commented on docs/src/main/sphinx/connector.rst:
```diff
@@ -31,6 +31,7 @@ from different data sources.
     Pinot           <connector/pinot>
```
removed doc related change from this pr.

MarvinCai commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java:
```diff
@@ -25,13 +25,16 @@
 import io.trino.spi.type.ArrayType;
```
Actually as we still have to duplicate some code in PulsarAvroColumnDecoder, I'll just keep these additional types in PulsarAvroColumnDecoder so existing connector won't be affected. Will only make method protected so they can be override.

MarvinCai replied:
Will create separate pr once change finalized.

MarvinCai commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java:
```diff
@@ -292,10 +302,20 @@ private static void serializePrimitive(BlockBuilder blockBuilder, Object value,
+        }
+
+        if (type instanceof TimestampType) {
+            type.writeLong(blockBuilder, (Long) value);
```
the change will be removed, while just for Pulsar connector it should be safe

MarvinCai commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/json/DefaultJsonFieldDecoder.java:
```diff
@@ -46,6 +46,13 @@
     private final long minValue;
     private final long maxValue;
 
+    protected DefaultJsonFieldDecoder()
```
it's a bit tricky, I'm trying to extend DefaultJsonFieldDecoder to reduce code duplication, while I do want to utilize the current constructor, but it can only invoke  `isSupportedType` in DefaultJsonFieldDecoder, which will prevent us from supporting new types, I'm adding this default constructor as workaround so we can skip `isSupportedType` when calling `super()`, any suggestion?

kokosing replied:
> any suggestion?

Yes, please undo ;)

Please make `PulsarJsonFieldDecoder` to implement `JsonFieldDecoder` without using `DefaultJsonFieldDecoder`.

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,219 @@
+        <air.main.basedir>${project.parent.basedir}</air.main.basedir>
+        <pulsar.version>2.9.0-SNAPSHOT</pulsar.version>
+        <bookkeeer.version>4.14.1</bookkeeer.version>
+        <netty.version>4.1.63.Final</netty.version>
```
assume this is at wrong line, but as we're using shaded jar now, don't need to directly include netty, will remove this.

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,219 @@
+        <dependency>
+            <groupId>org.testng</groupId>
+            <artifactId>testng</artifactId>
+            <version>6.10</version>
```
updated.

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,219 @@
+            <scope>test</scope>
+        </dependency>
+
+
```
fixed.

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,219 @@
+        <dependency>
+            <groupId>org.apache.avro</groupId>
+            <artifactId>avro</artifactId>
+            <version>1.9.2</version>
```
removed.

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,219 @@
+        <dependency>
+            <groupId>com.google.inject</groupId>
+            <artifactId>guice</artifactId>
+            <version>5.0.1</version>
```
removed

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,219 @@
+            <artifactId>pulsar-client-mledger-shaded</artifactId>
+            <version>${pulsar.version}</version>
+            <exclusions>
+                <exclusion>
```
the original Pulsar client has dependency on both javax & jakarta / bouncycastle.bcprov-ext-jdk15on & bouncycastle.bcprov-jdk15on which have duplicate classes, as they're not transient dependencies, can't exclude them with maven shade plugin, so have to exclude here.

kokosing replied:
duplicate and equal? Maybe you can relocate them (shade)?

MarvinCai commented on plugin/trino-pulsar/pom.xml:
```diff
@@ -0,0 +1,219 @@
+
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
```
accidentally using slf4j in couple places, changed to airlift:log and removed this dep

MarvinCai commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorId.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Unique identifier of a connector.
+ */
+public class PulsarConnectorId
```
sure, switched to CatalogName.

MarvinCai commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorModule.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * A wrapper to deserialize the Trino types.
+     */
+    public static final class TypeDeserializer
```
done.

MarvinCai commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorManagedLedgerFactory.java:
```diff
@@ -0,0 +1,200 @@
+        metadataStore = MetadataStoreFactory.create(pulsarConnectorConfig.getZookeeperUri(),
+                MetadataStoreConfig.builder().build());
+        managedLedgerFactory = initManagedLedgerFactory(pulsarConnectorConfig);
+        statsProvider = PulsarConnectorUtils.createInstance(pulsarConnectorConfig.getStatsProvider(),
```
previously we want to able to load different statsprovider, but since we switched to JMX based implementation to align with project convention, we don't need it anymore, removed.

MarvinCai commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorMetricsTracker.java:
```diff
@@ -0,0 +1,376 @@
+import org.apache.bookkeeper.stats.NullStatsProvider;
+import org.apache.bookkeeper.stats.OpStatsLogger;
+import org.apache.bookkeeper.stats.StatsLogger;
+import org.apache.bookkeeper.stats.StatsProvider;
```
updated to use JMX based stats.

kokosing reviewed (commented):
Please let me know when the new changes are posted. I will take another look. Please rebase and squash commits.

MarvinCai commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorManagedLedgerFactory.java:
```diff
@@ -0,0 +1,200 @@
+                        offloadPolicies.getManagedLedgerOffloadDriver());
+            }
+            catch (IOException ioe) {
+                throw new RuntimeException("Failed to create offloader factory: ", ioe.getCause());
```
sure, not catching IOException, but in some place where method is inherited or in Lambda, still have to catch then re-throw.

MarvinCai commented:
@kokosing my bad, thought you might have more comments later
addressed your latest comments, will create separate pr for decoder changes once you think the change looks good.
also squashed commits, please take a look

kokosing commented:
> my bad, thought you might have more comments later

My bad. I thought will have more comments, but I didn't have time to take a look. Sorry.

kokosing commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java:
```diff
@@ -91,7 +91,7 @@ public AvroColumnDecoder(DecoderColumnHandle columnHandle)
         }
```
Why not to add support for short timestamp, date, time and real support to `AvroColumnDecoder`? Then we should be able to drop `PulsarAvroColumnDecoder`. WDYT?

MarvinCai replied:
The blocker is conversion from shaded version of avro record to original avro record so pulsar connector can utilize `AvroColumnDecoder`, as the whole connector use shaded version of deps.
I tried again and was able to make the conversion this time.
I also move the support of additional type to `AvroColumnDecoder`, but I'm not sure how will this affect existing connectors? My understanding is it shouldn't be problem as it's "additional" suuport.

tisonkun replied:
> Why not to add support for short timestamp, date, time and real support to `AvroColumnDecoder`? Then we should be able to drop `PulsarAvroColumnDecoder`. WDYT?

I like this idea. @kokosing is there any existing issue? Or I can try to drive one.

tisonkun replied:
FYI https://github.com/trinodb/trino/issues/13069.

BTW pulsar-sql supports Decimal type in:

* https://github.com/apache/pulsar/commit/6b004ed6a2554ab826a00aa2a177963de3c5f44b
* https://github.com/apache/pulsar/commit/0c6e2ca24fd368d11a769233bb2041f6cc4a8374

It's possible to add them to AvroColumnDecoder later.

Also I think that this is not a blocker to add pulsar plugin as we can duplicate code a bit for Pulsar decoders - they have their own logics.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarAdminClientProvider.java:
```diff
@@ -0,0 +1,54 @@
+    private PulsarAdminClientProvider()
+    { }
+
+    @NotNull
```
In Trino everything by default is not null. We annotate only `@Nullable` variables.

MarvinCai replied:
done

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnHandle.java:
```diff
@@ -0,0 +1,255 @@
+    private final Type type;
+
+    /**
+     * True if the column should be hidden.
```
Such comments are obvious, please drop them.

MarvinCai replied:
done

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnHandle.java:
```diff
@@ -0,0 +1,255 @@
+    @Override
+    public int hashCode()
+    {
+        int result = catalogName != null ? catalogName.hashCode() : 0;
```
Objects.hashCode, here and everywhere else

MarvinCai replied:
Regenerated with IntelliJ, now using Objects.hash or Objects.hashCode

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnHandle.java:
```diff
@@ -0,0 +1,255 @@
+            return false;
+        }
+
+        return Objects.equals(handleKeyValueType, that.handleKeyValueType);
```
It looks like equals is hand-written, can you regenerate it using Intellij?

MarvinCai replied:
done.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnHandle.java:
```diff
@@ -0,0 +1,255 @@
+    @Override
+    public String toString()
+    {
+        return "PulsarColumnHandle{"
```
Use `ToStringHelper`, here and everywhere else.

MarvinCai replied:
sure, updated.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarColumnMetadata.java:
```diff
@@ -0,0 +1,235 @@
+import java.util.Objects;
+
+public class PulsarColumnMetadata
+        extends ColumnMetadata
```
Please use compostion instead of inheritance

MarvinCai replied:
after closer look PulsarColumnMetadata is just carrying some additional properties, so I use properties of ColumnMetadata to carry such info when necessary and make PulsarColumnMetadata just a util class.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnector.java:
```diff
@@ -0,0 +1,87 @@
+    private final ConnectorMetadata metadata;
+    private final ConnectorSplitManager splitManager;
+    private final ConnectorRecordSetProvider recordSetProvider;
+    private final PulsarConnectorConfig pulsarConnectorConfig;
```
this variable is not used, please remove.

MarvinCai replied:
done.

kokosing reviewed (commented):
Few comments. Please focus now on decoders.

MarvinCai commented:
here's the pom for the shaded pulsar dep: https://github.com/MarvinCai/pulsar/blob/zxc/trino-module/pulsar-client-mledger-shaded/pom.xml

jiazhai commented:
@MarvinCai Seems there were some CI tests that failed, Would you please help take a look at it?

MarvinCai commented:
@jiazhai That's because we made a new maven module to provide a shaded jar for this connector, and we plan to add it under trinodb organization, and currently it's only available on my local dev environment so the ci task just fail to build this connector. Once we finalized the dependency we'll create that new module which should fix all the build failure.

nlu90 commented:
@kokosing @MarvinCai Just curious to learn what's the progress for this PR?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java:
```diff
@@ -0,0 +1,734 @@
+    @Override
+    public long getCompletedBytes()
+    {
+        return this.completedBytes;
```
remove `this.`

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java:
```diff
@@ -0,0 +1,734 @@
+                                            messageDeserialized.increment();
+                                        }
+                                        catch (InterruptedException e) {
+                                            //no-op
```
? We should not simply ignore interruption.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java:
```diff
@@ -0,0 +1,734 @@
+                    @Override
+                    public void accept(Entry entry)
+                    {
+                        try {
```
3x nested try. Can you please refactor this somehow?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java:
```diff
@@ -0,0 +1,734 @@
+        try {
+            currentMessage.release();
+        }
+        catch (IllegalReferenceCountException e) {
```
please do not swallow exceptions

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java:
```diff
@@ -0,0 +1,734 @@
+        else if (schemaName.equals(Schema.BYTEBUFFER.getSchemaInfo().getName())) {
+            return Schema.BYTEBUFFER.getSchemaInfo();
+        }
+        else {
```
remove `else`s here

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java:
```diff
@@ -0,0 +1,734 @@
+                cursor.close();
+            }
+            catch (Exception e) {
+                log.error(e);
```
throw?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarRecordCursor.java:
```diff
@@ -0,0 +1,734 @@
+            try {
+                cursor.close();
+            }
+            catch (Exception e) {
```
can we get interrupted exception here?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSplitManager.java:
```diff
@@ -0,0 +1,500 @@
+
+    private final PulsarConnectorCache pulsarConnectorManagedLedgerFactory;
+
+    private ObjectMapper objectMapper = new ObjectMapper();
```
JsonCodec?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSplitManager.java:
```diff
@@ -0,0 +1,500 @@
+public class PulsarSplitManager
+        implements ConnectorSplitManager
+{
+    private static final Logger log = Logger.get(PulsarSplitManager.class);
```
in other place you used `LOG`. Please use `log` everywhere.

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSplitManager.java:
```diff
@@ -0,0 +1,500 @@
+
+    private final String catalogName;
+
+    private final PulsarConnectorConfig pulsarConnectorConfig;
```
please do not store config as fields

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSplitManager.java:
```diff
@@ -0,0 +1,500 @@
+{
+    private static final Logger log = Logger.get(PulsarSplitManager.class);
+
+    private final String catalogName;
```
remove catalog name

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSplitManager.java:
```diff
@@ -0,0 +1,500 @@
+            }
+        }
+        catch (PulsarClientException e) {
+            throw new TrinoException(PULSAR_ADMIN_ERROR, "fail to create pulsar admin client", e);
```
TrinoException is user facing error. Please work  a bit more on error message so user can actually understand what happened. 

What is `PULSAR_ADMIN_ERROR`? When user gets this what they should do? Same for other error codes and please double check all places where you throw TrinoException

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarSplitManager.java:
```diff
@@ -0,0 +1,500 @@
+        try (PulsarAdmin pulsarAdmin = PulsarAdminClientProvider.getPulsarAdmin(pulsarConnectorConfig)) {
+            OffloadPoliciesImpl offloadPolicies = (OffloadPoliciesImpl) pulsarAdmin.namespaces()
+                                    .getOffloadPolicies(topicName.getNamespace());
+            if (offloadPolicies != null) {
```
What if it is null? Is it ok to pass it to `getSplitsPartitionedTopic`? Maybe rnn here?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarTableHandle.java:
```diff
@@ -0,0 +1,120 @@
+public class PulsarTableHandle
+        implements ConnectorTableHandle
+{
+    private final String catalogName;
```
remove?

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarTableHandle.java:
```diff
@@ -0,0 +1,120 @@
+    private final String schemaName;
+
+    /**
+     * The table name used by Trino.
```
remove such obvious comments

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/json/PulsarJsonFieldDecoder.java:
```diff
@@ -0,0 +1,438 @@
+        else if (type == BIGINT) {
+            return new AbstractMap.SimpleEntry((long) Long.MIN_VALUE, (long) Long.MAX_VALUE);
+        }
+        else {
```
remove `else` words in this method

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/json/PulsarJsonFieldDecoder.java:
```diff
@@ -0,0 +1,438 @@
+            this.value = value;
+            this.columnHandle = columnHandle;
+            this.minValue = minValue;
+            this.maxValue = maxValue;
```
rnn for all

kokosing commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/decoder/json/PulsarJsonFieldDecoder.java:
```diff
@@ -0,0 +1,438 @@
+                }
+            }
+            catch (NumberFormatException ignore) {
+                // ignore
```
please do not swallow exceptions

kokosing commented on plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestPulsarConnectorConfig.java:
```diff
@@ -0,0 +1,108 @@
+    {
+        PulsarConnectorConfig connectorConfig = new PulsarConnectorConfig();
+
+        final String managedLedgerOffloadDriver = "s3";
```
remove `final`

kokosing reviewed (commented):
Add you plugin to ./core/trino-server/src/main/provisio/presto.xml

It would be nice to have product test for this to make sure it actually works end to end.

lhotari commented:
@MarvinCai Any ETA for getting the PR review comments addressed?

lhotari commented:
@MarvinCai Are you still actively working on this PR? Do you need any help in getting this to completion?
/cc @sijie @merlimat @codelipenghui

tisonkun commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarAdminClientProvider.java:
```diff
@@ -0,0 +1,52 @@
+    private PulsarAdminClientProvider()
+    { }
+
+    public static PulsarAdmin getPulsarAdmin(PulsarConnectorConfig config) throws PulsarClientException
```
Do we have a style checker for all these requirements, or even some tasks like spotless to apply these rules automatically?

tisonkun commented:
I'll take a look into this patch and try to take it over in this month if @MarvinCai doesn't make new progresses. However, it seems almost a rewrite of pulsar-presto-plugin released from within the pulsar community. Even I cannot run the Pulsar SQL example from the nightly version https://github.com/apache/pulsar/issues/16354.

Another possible approach is we bump the dependency from "Presto" to "Trino" and change the brands everywhere. After we make this change, we push the pulsar-trino-plugin to this repository and perhaps it can provide a more smooth experience - this patch is made before Pulsar released 2.8.0 but now it's 2.11.0. We may collaborate on the pulsar repo to make the existing plugin a good fit to trino while keep it synced with pulsar release so that we don't diverge.

What do you think? cc @MarvinCai @kokosing @hashhar @lhotari

tisonkun commented on plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarHandleResolver.java:
```diff
@@ -0,0 +1,89 @@
+/**
+ * This class helps to resolve classes for the Trino Pulsar connector.
+ */
+public class PulsarHandleResolver
```
API changed in https://github.com/trinodb/trino/commit/9c2f86d941f2adebb8db94fdb3382b11f02f25ce.

tisonkun commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java:
```diff
@@ -25,13 +25,16 @@
 import io.trino.spi.type.ArrayType;
 import io.trino.spi.type.BigintType;
```
I saw that Kafka plugin writes its own `AvroRowEncoder` and actually Pulsar writes its own `AvroColumnDecoder ` in `pulsar-sql` now. The major concern can be since `PulsarAvroColumnDecoder` is derive from `AvroColumnDecoder`, we may want to keep their logic synced. All significant fields and methods in `AvroColumnDecoder` seems `private` though.

tisonkun commented:
I'm starting upgrade presto to trino in pulsar codebase at branch https://github.com/tisonkun/pulsar/tree/presto-to-trino.

It seems we meet four major breaking changes:

1. `ConnectorHandleResolver` is removed.
2. `ConnectorTableLayoutHandle` and `ConnectorTableLayoutResult` are removed.
3. `ColumnMetadata` is changed so that we cannot `extends` it.
4. `Decimal` is breaking changed in https://github.com/trinodb/trino/commit/fb88c0bf589db55d77223fc31aab29acbb2d0698.

cc @kokosing @MarvinCai @lhotari perhaps we can bump presto to trino in pulsar codebase first so that we can deliver the change to users quickly and later after we switch to trino dependency it will be smoother to contribute to trino.

I hope that we can collaborate on the branch posted here and looking forward to your ideas.

colebow commented:
:wave: @MarvinCai - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

@tisonkun - if you're working on running with this, please reach out if you need help with anything.

tisonkun commented:
@colebow Thanks for your follow-up!

I have a pending PR https://github.com/trinodb/trino/pull/13070 that needs more feedback. If you guys think it may not be a common use case, I can integrate it into the trino-pulsar-plugin level.

For continuing this contribution, there're still other efforts that should be done to catch up with the latest Trino version. As listed above, a few breaking changes are made and the trino version we're using now is 363 (compared to the latest one 400).
